### ISSUE TITLE
small assistance with random seeding

### ISSUE THREAD

**Issue by aleqx** [fireice-uk/xmr-stak#1095]
tl;dr - I'd be very grateful if you could point me to the file+line in the source where I can change the search from a fixed starting nonce to a random starting nonce? For the CPU miner, is it:

https://github.com/fireice-uk/xmr-stak/blob/2ae7260b90fe3dbe835ba2489519510f0e57d770/xmrstak/backend/cpu/minethd.cpp#L397

https://github.com/fireice-uk/xmr-stak/blob/2ae7260b90fe3dbe835ba2489519510f0e57d770/xmrstak/backend/cpu/minethd.cpp#L552

https://github.com/fireice-uk/xmr-stak/blob/2ae7260b90fe3dbe835ba2489519510f0e57d770/xmrstak/backend/cpu/minethd.cpp#L580

?

not-tl;dr: I'm writing a proxy that works with Nicehash. Nicehash is nasty as they strip most useful nonce info in the login reply, so there is no way to isolate miners based on nonce space, and they clash, sending duplicate solutions. Specifically: when multiple xmr-stak miners are connected to the proxy, they all receive the same share, at the same time, and all compute the same solution, then all submit that solution to Nicehash, which accepts the first one and rejects all others as duplicates. This is because xmr-stak starts searching for a solution from the same nonce every time, hence they will all clash all the time. 

The/A fix is to start from a random nonce (in lack of a way to tell each miner to start from a different nonce). I had the same problem with nheqminer, which I patched so it starts from a random nonce. I was more familiar with the nheqminer code than xmr-stak, so I'd appreciate a pointer to achieve that.

**Comment by fireice-uk** [comment]
It is not a good way. Please look into https://en.wikipedia.org/wiki/Birthday_problem - chances of miners colliding by pure accident are much greater than 1 / 2^32

**Comment by aleqx** [comment]
Thanks for replying. I never said collision probability is 2^-32 and I'm quite well aware of the birthday paradox (collision probability for 32 bit integers is 1e-6 for 93 nodes, 1e-5 for 293 nodes, 1e-4 for 927 nodes ... totally negligible considering stale share probability is order of magnitudes higher. You'd need 9200+ nodes to approach 1% collision probability).

But your answer doesn't address my question at all! :) I was asking for code assistance. Where in the code could I randomize the seed/nonce and are there any non-collision related aspects I need to be aware of in Cryptonight?

p.s. Collisions happen 100% of the time without randomizing (and I can't have nodes talk to each other or to a central entity to issue unique nonces), so ...

**Comment by psychocrypt** [comment]
check the files `minethd.cpp`.
Never the less the pool need to avoid that two miner calculate on the same
hash.

**Comment by aleqx** [comment]
I was hoping for some more detailed help than that (as you can tell from my first post, I've already identified the file and candidate lines).

Pools already reject duplicate shares, so that's not an issue. Or did you mean something else?

**Comment by aleqx** [comment]
@fireice-uk I see you edited your answer and pointed to a code entry - thanks. I'm not worried about multiple pools (there will be a single pool in this case). Do I still need to mess with executor.cpp and `xmrstak::pool_data` rather than just change piNonce in minethd.cpp?

**Comment by psychocrypt** [comment]
Check the file globalStats.hpp You need to change one line.

**Comment by JerichoJones** [comment]
@aleqx Did you get what you needed? If so, please close.

**Comment by JerichoJones** [comment]
Closing due to lack of response.
